models/User.php

```diff
@@ -145,22 +160,95 @@ public function validateAuthKey($authKey)
      */
     public function validatePassword($password)
     {
-        return $this->password === $password;
+        return Yii::$app->security->validatePassword($password, $this->password_hash);
+    }
+
+    public function login() {
+        $this->token = $this->generateToken();
+        return $this->save();
+    }
+
+    /**
+     * @return bool
+     * @throws \yii\base\Exception
+     */
+    public function register() {
+        $this->password_hash = Yii::$app->security->generatePasswordHash($this->password);
+        $this->token = $this->generateToken();
+        return $this->save();
+    }
+
+    /**
+     * @return string
+     * @throws InvalidConfigException
+     */
+    protected function generateToken() {
+        $signer = new Sha256();
+
+        if (!isset(\Yii::$app->jwt)) {
+            throw new InvalidConfigException('You must setup JWT component');
+        }
+
+        $token = \Yii::$app->jwt->getBuilder()
+            ->setExpiration(time() + self::TOKEN_EXPIRATION_TIMEOUT)
+            ->set('user_id', $this->id)
+            ->sign($signer, \Yii::$app->jwt->key)
+            ->getToken();
+        return strval($token);
     }
 
     /**
      * @return \yii\db\ActiveQuery
      */
     public function getFollows()
     {
-        return $this->hasMany(Follow::className(), ['follow_id' => 'id']);
+        return $this->hasMany(Follow::class, ['follow_id' => 'id']);
```

Comment:
`::className()` is in Yii2 convention 

See the guide https://www.yiiframework.com/doc/guide/2.0/en/db-active-record#declaring-relations

Reply:
className deprecated in Yii2 since Yii2 2.0.14. See:
Chg #15633: Deprecated yii\base\BaseObject::className() in favor of native PHP syntax ::class, which does not trigger autoloading (brandonkelly)
in https://github.com/yiisoft/yii2/blob/master/framework/CHANGELOG.md

Reply:
okay than we need a guide update.